File: AnagramLister/AnagramLister.py
from pyspark import SparkContext, SparkConf
import sys

def appendWord(x, y):
	if x == y:
		return x
	else:
		return x + ',' + y

def removeDuplicates(x):
	str_array = x[1].split(',')
	str_set = set(str_array)
	return [x[0], str_set]

def length(x):
	arr = x.split(',')
	return len(arr)

inputs = sys.argv[1]
outputs = sys.argv[2]

conf = SparkConf().setAppName('anagram lister')
conf.set("spark.hadoop.validateOutputSpecs", "false")
context = SparkContext(conf=conf)
context._jsc.hadoopConfiguration().set("textinputformat.record.delimiter", " ")

rdd = context.textFile(inputs)\
			 .filter(lambda x:x!=' ')\
			 .map(lambda x:x.encode("utf-8"))\
			 .map(lambda x:x.lower())\
			 .map(lambda x:(''.join(sorted(x)), x))\
			 .reduceByKey(lambda x,y: appendWord(x,y))\
			 .map(lambda x:removeDuplicates(x))\
			 .filter(lambda x:len(x[1])>=2)\
			 .coalesce(1)\
			 .saveAsTextFile(outputs)

File: Scala/AnagramLister/AnagramLister.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

object AnagramLister	{
	def reduceDuplicates(args: (String, String)) :(String, Array[String]) = {
		val key = args._1
		val value = args._2

		val valueSplitted = value.split(",")
		val valueSet = valueSplitted.toSet
		val valueArray = valueSet.toArray

		(key, valueArray)
	}
	

	def main(args: Array[String])	{
		val inputFile = args(0)
		val outputFile = args(1)
		val conf = new SparkConf().setAppName("anagram lister")		
		val sc = new SparkContext(conf)

		val inputData = sc.textFile(inputFile)
		val rdd = inputData.flatMap(line => line.split(" "))
		val anagramMap = rdd.map(x => (x.sorted, x))
		val anagramMapTuple = anagramMap.reduceByKey((x,y) => x+","+y)
		val reduceDup = anagramMapTuple.map(x => reduceDuplicates(x))
		val result = reduceDup.filter(x => x._2.length > 1)

		result.saveAsTextFile(outputFile)
	}
}